app/experimenter/nimbus-ui/src/components/FormBranches/reducer.tsx

```diff
@@ -0,0 +1,266 @@
+> & {
+  referenceBranch: null | AnnotatedBranch;
+  treatmentBranches: null | AnnotatedBranch[];
+  equalRatio: boolean;
+  lastId: number;
+};
+
+export type FormBranchesSaveState = Pick<
+  getExperiment_experimentBySlug,
+  "featureConfig" | "referenceBranch" | "treatmentBranches"
+>;
+
+export type FormBranchesAction =
+  | AddBranchAction
+  | RemoveBranchAction
+  | UpdateBranchAction
+  | RemoveFeatureConfigAction
+  | SetEqualRatioAction
+  | SetFeatureConfigAction;
+
+type AddBranchAction = {
+  type: "addBranch";
+};
+
+function addBranch(state: FormBranchesState) {
+  let { lastId, referenceBranch, treatmentBranches } = state;
+  lastId++;
+
+  if (referenceBranch === null) {
+    referenceBranch = makeBranch(lastId, "New control");
+  } else {
+    treatmentBranches = [
+      ...(treatmentBranches || []),
+      makeBranch(state.lastId, `New treatment ${lastId}`),
+    ];
+  }
+
+  return {
+    ...state,
+    lastId,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+type RemoveBranchAction = {
+  type: "removeBranch";
+  idx: number;
+};
+
+function removeBranch(state: FormBranchesState, { idx }: RemoveBranchAction) {
+  const treatmentBranches = state.treatmentBranches || [];
+  return {
+    ...state,
+    treatmentBranches: [
+      ...treatmentBranches.slice(0, idx),
+      ...treatmentBranches.slice(idx + 1),
+    ],
+  };
+}
+
+type UpdateBranchAction = {
+  type: "updateBranch";
+  idx: number;
+  value: AnnotatedBranch;
+};
+
+function updateBranch(
+  state: FormBranchesState,
+  { idx, value }: UpdateBranchAction,
+) {
+  let { referenceBranch, treatmentBranches } = state;
+  const updatedBranch = { ...value };
+  if (idx === REFERENCE_BRANCH_IDX) {
+    referenceBranch = updatedBranch;
+  } else {
+    treatmentBranches = (treatmentBranches || []).slice();
+    treatmentBranches[idx] = updatedBranch;
+  }
+  return {
+    ...state,
+    referenceBranch,
+    treatmentBranches,
+  };
+}
+
+type RemoveFeatureConfigAction = {
+  type: "removeFeatureConfig";
+};
+
+function removeFeatureConfig(state: FormBranchesState) {
+  return {
+    ...state,
+    featureConfig: null,
+  };
+}
+
+type SetFeatureConfigAction = {
+  type: "setFeatureConfig";
+  value: FormBranchesState["featureConfig"];
+};
+
+function setFeatureConfig(
+  state: FormBranchesState,
+  { value: featureConfig }: SetFeatureConfigAction,
+) {
+  return {
+    ...state,
+    featureConfig,
+  };
+}
+
+type SetEqualRatioAction = {
+  type: "setEqualRatio";
+  value: FormBranchesState["equalRatio"];
+};
+
+function setEqualRatio(
+  state: FormBranchesState,
+  { value: equalRatio }: SetEqualRatioAction,
+) {
+  let { referenceBranch, treatmentBranches } = state;
+  if (equalRatio) {
+    if (referenceBranch !== null) {
+      referenceBranch = { ...referenceBranch, ratio: 1 };
+    }
+    if (Array.isArray(treatmentBranches)) {
+      treatmentBranches = treatmentBranches.map(
+        (treatmentBranch) =>
+          treatmentBranch && {
```

Comment:
What scenario might this be falsy? Does this map need a filter to ensure only branch objects make it through?

Reply:
No idea when it might be falsy - I don't think it will in practical use. The generated type has a null case, so this tries to cover that.